app/src/main/java/org/gdg/frisbee/android/gde/GdeActivity.java

```diff
@@ -94,102 +84,66 @@ public void networkFailure(Throwable error) {
+        HashMap<String, GdeList> gdeMap = extractCategoriesFromGdeList(directory);
+        List<GdeCategory> gdeCategoryList = convertCategoryMapToList(gdeMap);
+
+        mViewPagerAdapter = new GdeCategoryPagerAdapter(
+            getSupportFragmentManager(),
+            getString(R.string.about),
+            gdeCategoryList
+        );
+        mViewPager.setAdapter(mViewPagerAdapter);
+        mViewPager.addOnPageChangeListener(this);
+        mTabLayout.setupWithViewPager(mViewPager);
+
+        getAchievementActionHandler().handleLookingForExperts();
     }
 
-    public class GdeCategoryAdapter extends FragmentStatePagerAdapter implements ViewPager.OnPageChangeListener {
-        private final SparseArray<WeakReference<Fragment>> mFragments = new SparseArray<>();
-        private HashMap<String, GdeList> mGdeMap;
-
-        public GdeCategoryAdapter(FragmentManager fm) {
-            super(fm);
-            mGdeMap = new HashMap<>();
-        }
-
-        public void addMap(Map<String, GdeList> collection) {
-            mGdeMap.putAll(collection);
-        }
-
-        @Override
-        public int getCount() {
-            return mGdeMap.keySet().size() + 1;
-        }
-
-        @Override
-        public Fragment getItem(int position) {
-            if (position == 0) {
-                return PlainLayoutFragment.newInstance(R.layout.fragment_gde_about);
-            } else {
-                String key = mGdeMap.keySet().toArray(new String[mGdeMap.size()])[position - 1];
-                Fragment frag = GdeListFragment.newInstance(mGdeMap.get(key));
-                mFragments.append(position, new WeakReference<>(frag));
+    private HashMap<String, GdeList> extractCategoriesFromGdeList(GdeList directory) {
+        HashMap<String, GdeList> gdeMap = new HashMap<>();
 
-                return frag;
+        for (Gde gde : directory) {
+            if (gde.getProduct() == null) {
+                continue;
             }
-        }
+            for (String p : gde.getProduct()) {
+                String product = p.trim();
+                if (!gdeMap.containsKey(product)) {
+                    gdeMap.put(product, new GdeList());
+                }
 
-        @Override
-        public CharSequence getPageTitle(int position) {
-            if (position == 0) {
-                return getString(R.string.about);
-            } else if (position > -1 && position - 1 < mGdeMap.keySet().size()) {
-                String title = mGdeMap.keySet().toArray(new String[mGdeMap.size()])[position - 1];
-                title = title.length() > 14 ? Utils.getUppercaseLetters(title) : title;
-                return title;
-            } else {
-                return "";
+                gdeMap.get(product).add(gde);
             }
         }
+        return gdeMap;
+    }
 
-        @Override
-        public void onPageScrolled(int i, float v, int i2) {
+    private List<GdeCategory> convertCategoryMapToList(HashMap<String, GdeList> gdeMap) {
```

Comment:
Ordering of the GDEs are also wrong but I was more talking about the
ordering of the categories.  It's also random.

The adapter had the HashMap before. And it was trying to access the key by
position which is really bad.

On Tue, Jun 7, 2016, 17:00 Friedger Müffke <notifications@github.com> wrote:

> In app/src/main/java/org/gdg/frisbee/android/gde/GdeActivity.java
> <https://github.com/gdg-x/frisbee/pull/642#discussion_r66087848>:
>
> >
> > -        @Override
> > -        public void onPageScrolled(int i, float v, int i2) {
> > +    private List<GdeCategory> convertCategoryMapToList(HashMap<String, GdeList> gdeMap) {
>
> The ordering is the ordering of the list, isn't it? We get the hash map
> from the server. We could do some smarter conversions there. The adapter is
> using a list.
> … <#m_8970533723015702354_>
> On 7 Jun 2016 16:22, "Said Tahsin Dane" ***@***.***> wrote: In
> app/src/main/java/org/gdg/frisbee/android/gde/GdeActivity.java <#642
> (comment) <https://github.com/gdg-x/frisbee/pull/642#discussion_r66079786>>:
> > > - @Override <https://github.com/Override> > - public void
> onPageScrolled(int i, float v, int i2) { > + private List<GdeCategory>
> convertCategoryMapToList(HashMap<String, GdeList> gdeMap) { Using HashMap
> in the adapter is also really really bad. I was wondering about the
> ordering of the categories in the ViewPager titles. The ordering was
> random. I think that is also because of this HashMap. I can move this
> convertion code to a background thread. — You are receiving this because
> you commented. Reply to this email directly, view it on GitHub <
> https://github.com/gdg-x/frisbee/pull/642/files/0f06e25e4d2c76b0917f7785a4d78b91c383f043#r66079786>,
> or mute the thread <
> https://github.com/notifications/unsubscribe/ABYcWZ7h200nqS2qzdOPUiqRrpulovKiks5qJX6JgaJpZM4IuTJd>
> .
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gdg-x/frisbee/pull/642/files/0f06e25e4d2c76b0917f7785a4d78b91c383f043#r66087848>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAulyxIlKo2qXHIrM-uhXuYGAOCWrKthks5qJYepgaJpZM4IuTJd>
> .
>